nexus/db-queries/src/db/datastore/clickhouse_policy.rs

```diff
@@ -0,0 +1,267 @@
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?
+        };
+
+        match num_inserted {
+            0 => Err(Error::invalid_request(format!(
+                "policy version {} is not the most recent",
+                policy.version
+            ))),
+            1 => Ok(()),
+            // This is impossible because we are explicitly inserting only one
+            // row with a unique primary key.
+            _ => unreachable!("query inserted more than one row"),
+        }
+    }
+
+    /// Insert the first clickhouse policy in the database at version 1.
+    ///
+    /// Only insert this policy if no other policy exists yet.
+    ///
+    /// Return the number of inserted rows or an error.
+    async fn clickhouse_policy_insert_first_policy(
+        &self,
+        opctx: &OpContext,
+        policy: &ClickhousePolicy,
+    ) -> Result<usize, Error> {
+        sql_query(
+            r"INSERT INTO clickhouse_policy 
+              (version, clickhouse_mode, clickhouse_cluster_target_servers, 
+               clickhouse_cluster_target_keepers, time_created) 
+             SELECT $1, $2, $3, $4, $5 
+             WHERE NOT EXISTS (SELECT * FROM clickhouse_policy)",
+        )
+        .bind::<sql_types::BigInt, SqlU32>(policy.version.into())
+        .bind::<ClickhouseModeEnum, DbClickhouseMode>((&policy.mode).into())
+        .bind::<sql_types::SmallInt, SqlU8>(policy.mode.target_servers().into())
+        .bind::<sql_types::SmallInt, SqlU8>(policy.mode.target_keepers().into())
+        .bind::<sql_types::Timestamptz, _>(policy.time_created)
+        .execute_async(&*self.pool_connection_authorized(opctx).await?)
+        .await
+        .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::db::datastore::test_utils::datastore_test;
+    use nexus_inventory::now_db_precision;
+    use nexus_test_utils::db::test_setup_database;
+    use nexus_types::deployment::ClickhouseMode;
+    use omicron_test_utils::dev;
+
+    #[tokio::test]
+    async fn test_clickhouse_policy_basic() {
+        // Setup
+        let logctx = dev::test_setup_log("test_clickhouse_policy_basic");
+        let mut db = test_setup_database(&logctx.log).await;
+        let (opctx, datastore) = datastore_test(&logctx, &db).await;
+
+        // Listing an empty table should return an empty vec
+
+        assert!(datastore
+            .clickhouse_policy_list(&opctx, &DataPageParams::max_page())
+            .await
+            .unwrap()
+            .is_empty());
+
+        // Fail to insert a policy with version 0
+        let mut policy = ClickhousePolicy {
+            version: 0,
+            mode: ClickhouseMode::SingleNodeOnly,
+            time_created: now_db_precision(),
+        };
+
+        assert!(datastore
+            .clickhouse_policy_insert_latest_version(&opctx, &policy)
+            .await
+            .is_err());
```

Comment:
Nit - maybe check the errors here and make sure they're what you expect? `err.to_string().contains("blah blah")` or somesuch

Reply:
These are all CRDB specific, and without triggering failure I'm not even sure what they should be.

Reply:
Well, maybe at least check the couple cases where these methods are supposed to return specific errors that are added in this PR (`policy version must be greater than 0` and `policy version {} is not the most recent`)?

Reply:
Totally reasonable! Not sure where my head was on this. I was mostly thinking about errors related to bad sql parsing that I got unexpectedly early on in testing.

Done in [556ccb7](https://github.com/oxidecomputer/omicron/pull/6862/commits/556ccb7ca42da8f6597ecdc2776cfc2e9c58f8b6)